mujoco/src/sim.rs

```diff
@@ -0,0 +1,162 @@
+use crate::{
+    helpers::{extract_vector_float_f64, Local},
+    Model, State,
+};
+
+pub struct Simulation {
+    pub state: State,
+    pub model: Model,
+}
+
+impl Simulation {
+    pub fn new(model: Model) -> Self {
+        let state = State::new(&model);
+        Self { state, model }
+    }
+
+    /// Set control vector
+    pub fn control(&self, control: &[f64]) {
+        let mj_data = self.state.ptr;
+        let raw_vec = unsafe { (*mj_data).ctrl };
+
+        if control.len() != self.model.nu() {
+            return;
+        }
+
+        #[allow(clippy::needless_range_loop)]
+        for i in 0..self.model.nu() {
+            unsafe {
+                *raw_vec.add(i) = control[i];
+            }
+        }
+    }
+
+    /// Evaulate constraint forces ans sensors
+    pub fn evaluate_sensors(&self) {
+        unsafe {
+            mujoco_rs_sys::no_render::mj_sensorPos(self.model.ptr(), self.state.ptr());
+            mujoco_rs_sys::no_render::mj_sensorVel(self.model.ptr(), self.state.ptr());
+            mujoco_rs_sys::no_render::mj_sensorAcc(self.model.ptr(), self.state.ptr());
+        };
+    }
+
+    /// Advance simulation by one step
+    pub fn step(&self) {
+        unsafe {
+            mujoco_rs_sys::no_render::mj_step(self.model.ptr(), self.state.ptr());
+        };
+    }
+
+    /// Returns positions of bodies in inertial frame
+    pub fn xpos(&self) -> Vec<[f64; 3]> {
+        let mj_data = self.state.ptr();
+        let raw_vec = unsafe { (*mj_data).xpos };
+
+        let raw_xpos =
+            extract_vector_float_f64(raw_vec as *mut Local<f64>, 3, self.model.ngeom());
+
+        let mut xpos: Vec<[f64; 3]> = Vec::new();
+
+        for i in 0..self.model.ngeom() {
+            let entry: [f64; 3] =
+                [raw_xpos[i * 3], raw_xpos[i * 3 + 1], raw_xpos[i * 3 + 2]];
+            xpos.push(entry);
+        }
+
+        xpos
+    }
+
+    /// Returns rotations of bodies
+    pub fn xquat(&self) -> Vec<[f64; 4]> {
+        let mj_data = self.state.ptr();
+        let raw_vec = unsafe { (*mj_data).xquat };
+        let raw_quat =
```

Comment:
This is a bit of a nit, but wouldn't it be better for readability to use a proper Quaternion type?

Reply:
Which quat implementation? 

https://github.com/PistonDevelopers/quaternion — small footprint
or perhaps https://docs.rs/nalgebra/latest/nalgebra/?search=Quaternion

Reply:
I was thinking nalgebra's, but really any is fine.